### PR TITLE
authority: add jwk updater to authority server

### PR DESCRIPTION
**Pull request by joyqvq**
## Description 

cache a jwk_as_bytes in authority server with default value. server spawns a a process that fetch google jwk from endpoint every hour. aux_verify_data is passed from authority state to verify tx sig in envelope. if jwk_as_byte is none, it means its called from consensus verifier and hence no need to verify. 

zk login authenticator added in a later pr will use this jwk
## Test Plan 

unit test and local ran validators. 
---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-types/src/message_envelope.rs [hunk 1]
```diff
@@ -224,7 +226,6 @@ where
     // TODO: Eventually we should remove all calls to verify_signature
     // and make sure they all call verify to avoid repeated verifications.
     pub fn verify_signature(&self, committee: &Committee) -> SuiResult {
-        self.data.verify(Some(self.auth_sig().epoch))?;
```
Comment (mystenmark) [inline crates/sui-types/src/message_envelope.rs:227]: why was this removed?

Comment (joyqvq) [reply crates/sui-types/src/message_envelope.rs:227]: sorry that meant to be highlighted instead of deleted. this method is invoked at many places that i dont have access to state.jwk_as_bytes, most of them are in the client side (safe client, authority aggregator). do we need a zk login tx pass this verification or its ok to skip?

File: crates/sui-core/src/authority.rs [hunk 2]
```diff
@@ -569,6 +599,54 @@ impl AuthorityState {
...
+                // Sleep for 1 hour
+                tokio::time::sleep(Duration::from_secs(3600)).await;
+            }
+        });
+    }
+
+    async fn update_google_jwk(&self) -> Result<(), SuiError> {
+        let client = Client::new();
+        let response = client
+            .get("https://www.googleapis.com/oauth2/v2/certs")
```
Comment (mystenmark) [inline crates/sui-core/src/authority.rs:627]: i think you will probably have to bypass doing an actual http fetch in order for this to work with simtest. but it may not matter until you add specific zklogin tests.

File: crates/sui-core/src/authority.rs [hunk 3]
```diff
@@ -569,6 +599,54 @@ impl AuthorityState {
...
+    }
+
+    pub fn start_jwk_updater(self: Arc<Self>) {
+        debug!("Starting JWK updater thread for authority");
+        tokio::task::spawn(async move {
+            loop {
+                // Update the JWK value in the authority server
+                let res = self.update_google_jwk().await;
+                if let Err(e) = res {
+                    debug!("Error when fetching JWK {:?}", e);
```
Comment (mystenmark) [inline crates/sui-core/src/authority.rs:616]: this should be `warn!`

File: crates/sui-core/src/authority_server.rs [hunk 4]
```diff
@@ -332,7 +332,7 @@ impl ValidatorService {
         let tx_verif_metrics_guard = metrics.tx_verification_latency.start_timer();
         let transaction = epoch_store
             .signature_verifier
-            .verify_tx(transaction.data())
+            .verify_tx(transaction.data(), Some(state.get_google_jwk_as_bytes()))
```
Comment (mystenmark) [inline crates/sui-core/src/authority_server.rs:335]: can we instead push the new jwk to the signature verifier whenever it changes, e.g.

```
signature_verifier.update_jwk(new_jwk)
```

File: crates/sui-types/src/message_envelope.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-types/src/message_envelope.rs:227]: I don't think we should skip verification entirely on the client. It would probably be best if clients fetched the JWK themselves, which should be quite simple to do. I think its okay if they aren't sync with the validators though, they can just fetch directly from google.

The reasoning here is that zklogin verification is quite a bit more expensive than previous authenticators, and is potentially a DoS attack surface. So we may need to rate limit or ban clients that send invalid zklogins. But the "client" from the validators point of view is usually a fullnode. So this in turn means that fullnodes would need to ensure sigs are valid before sending on to the validators, or end users could get the fullnodes banned.

File: crates/sui-types/src/signature.rs [hunk 5]
```diff
@@ -31,6 +48,7 @@ pub trait AuthenticatorTrait {
 /// This way MultiSig (and future Authenticators) can implement its own `verify`.
 #[enum_dispatch(AuthenticatorTrait)]
 #[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash)]
+#[allow(clippy::large_enum_variant)]
```
Comment (mystenmark) [inline crates/sui-types/src/signature.rs:51]: this doesn't look like it has any large variants? is this still needed?

File: crates/sui-core/src/authority_server.rs (hunk 4 above)
Comment (joyqvq) [reply crates/sui-core/src/authority_server.rs:335]: you are suggesting to put the jwk updater inside AuthorityPerEpochStore (that has the signature_verifier) instead of AuthorityState?

File: crates/sui-types/src/signature.rs (hunk 5 above)
Comment (joyqvq) [reply crates/sui-types/src/signature.rs:51]: multisig is quite large and triggered this

File: crates/sui-core/src/authority.rs (hunk 2 above)
Comment (joyqvq) [reply crates/sui-core/src/authority.rs:627]: makes sense. i can just use a cfg(test) vs cfg(not(test)) for simtest to use a dummy response right?

**Comment by joyqvq** [comment]
https://github.com/MystenLabs/sui/pull/12273 in favor of this approach instead
